tools/codegen/api/python.py

```diff
@@ -0,0 +1,955 @@
+            else:
+                typename = f'{typename}[{size}]'
+
+        # add default
+        if self.default is not None:
+            default = {
+                'nullptr': 'None',
+                'c10::nullopt': 'None',
+                '{}': 'None',
+            }.get(self.default, self.default)
+            return f'{typename} {name}={default}'
+        else:
+            return f'{typename} {name}'
+
+    @property
+    def size(self) -> Optional[int]:
+        l = self.type.is_list_like()
+        return l.size \
+            if l is not None and l.size is not None and str(l.elem) != 'bool' else None
+
+
+@dataclass(frozen=True)
+class PythonOutArgument(PythonArgument):
+    # In Python signature multiple output fields are packed into one 'out' argument.
+    # When binding to C++, it's first binded to a local 'out' variable:
+    #   'auto out = _r.tensorlist_n<2>(2);',
+    # then binded to scattered C++ output arguments as 'out[0]', 'out[1]', and etc.
+    # TODO: maybe don't need keep scattered out fields for python signature?
+    outputs: Tuple[PythonArgument, ...]
+
+    @staticmethod
+    def from_outputs(outputs: Tuple[PythonArgument, ...]) -> Optional['PythonOutArgument']:
+        if not outputs:
+            return None
+
+        size = len(outputs)
+        if size == 1:
+            return PythonOutArgument(
+                name=outputs[0].name,
+                type=outputs[0].type,
+                cpp_type_str=outputs[0].cpp_type_str,
+                default='None',
+                default_init=None,
+                outputs=outputs,
+            )
+        elif size > 1:
+            return PythonOutArgument(
+                name='out',
+                type=ListType(BaseType(BaseTy.Tensor), size),
+                cpp_type_str='TensorList',
+                default='None',
+                default_init=None,
+                outputs=outputs,
+            )
+        raise AssertionError(r'Unexpected PythonOutArgument size')
+
+
+@dataclass(frozen=True)
+class PythonSignature:
+    # Base operator name, without inplace/outplace suffix.
+    name: str
+
+    # Positional arguments.
+    # TODO: create a dedicated SelfArgument type for 'self'?
+    input_args: Tuple[PythonArgument, ...]
+
+    # Keyword arguments excluding the 'out' argument and scattered kwargs belonging
+    # to TensorOptions (dtype, layout, device, pin_memory, requires_grad, etc).
+    input_kwargs: Tuple[PythonArgument, ...]
+
+    output_args: Optional[PythonOutArgument]
+
+    # These are scattered kwargs arguments belonging to TensorOptions.
+    # When binding to C++, they are packed into a TensorOptions object 'options'.
+    # It's possible that the C++ signature doesn't take TensorOptions object (e.g.
+    # for out variant), in which case they will be used as scattered fields without
+    # being packed into 'options'.
+    # TODO: maybe create a PythonTensorOptionsArgument?
+    tensor_options_args: Tuple[PythonArgument, ...]
```

Comment:
should this be Optional, since not all ops take in tensor_options? Or I guess an empty tuple represents lack of a tensor_options argument

Reply:
Right - right now empty means no tensor options... I'm fine if you think explicit null value makes more sense.
Note that this field doesn't only have all v.s. nothing states - you can have some tensor options fields in python signature but not all of them, e.g. 'requires_grad'. It's why I feel less guilty of not differentiating empty v.s. null...

Reply:
@bhosmer after reading this PR, I am pretty convinced that grouping (`group_arguments`) should be incorporated directly in the model, and I'm also feeling more bullish on your proposal at https://github.com/pytorch/pytorch/pull/45890#discussion_r503646917 which would save Jiakai from a lot of manual dissecting here (so we should do something like indexes (blegh) to retain order correctly). The big difference is that Jiakai has to do more reordering that we didn't need in the old codegen; whenever reordering is involved, being able to explicitly access the pieces that are being reordered is a big win.